src/rust/inetstack/protocols/layer3/icmpv4/peer.rs

```diff
@@ -165,22 +174,12 @@ impl SharedIcmpv4Peer {
             debug!("reply ping ({}, {}, {})", dst_ipv4_addr, id, seq_num);
             // Send reply message.
             let local_ipv4_addr: Ipv4Addr = self.local_ipv4_addr;
-            let mut message: Icmpv4Message = match Icmpv4Message::new(
-                Ipv4Header::new(local_ipv4_addr, dst_ipv4_addr, IpProtocol::ICMPv4),
-                Icmpv4Header::new(Icmpv4Type2::EchoReply { id, seq_num }, 0),
-                data,
-            ) {
-                Ok(msg) => msg,
-                Err(e) => {
-                    warn!("Could not assemble packet: {:?}", e);
-                    continue;
-                },
-            };
+            let icmp_hdr: Icmpv4Header = Icmpv4Header::new(Icmpv4Type2::EchoReply { id, seq_num }, 0);
+            icmp_hdr.serialize_and_attach(&mut buf);
+            let ipv4_hdr: Ipv4Header = Ipv4Header::new(local_ipv4_addr, dst_ipv4_addr, IpProtocol::ICMPv4);
```

Comment:
Not really sure how to get the ICMP packet through the layer 3 IP layer. The OSI model clearly says that it belongs at layer 3 because it doesn't understand ports, but we need to think about how to put it still above the IP layer.